BTCPayServer/HostedServices/DbMigrationsHostedService.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading.Tasks;
+using BTCPayServer.Data;
+using BTCPayServer.Services;
+using BTCPayServer.Services.Invoices;
+
+namespace BTCPayServer.HostedServices
+{
+    /// <summary>
+    /// In charge of all long running db migrations that we can't execute on startup in MigrationStartupTask
+    /// </summary>
+    public class DbMigrationsHostedService : BaseAsyncService
+    {
+        private readonly InvoiceRepository _invoiceRepository;
+        private readonly SettingsRepository _settingsRepository;
+        private readonly ApplicationDbContextFactory _dbContextFactory;
+
+        public DbMigrationsHostedService(InvoiceRepository invoiceRepository, SettingsRepository settingsRepository, ApplicationDbContextFactory dbContextFactory)
+        {
+            _invoiceRepository = invoiceRepository;
+            _settingsRepository = settingsRepository;
+            _dbContextFactory = dbContextFactory;
+        }
+        
+
+        internal override Task[] InitializeTasks()
+        {
+            return new Task[] { ProcessMigration() };
+        }
+
+        protected async Task ProcessMigration()
+        {
+            var settings = await _settingsRepository.GetSettingAsync<MigrationSettings>();
+            if (settings.MigratedInvoiceTextSearchPages != int.MaxValue)
+            {
+                await MigratedInvoiceTextSearchToDb(settings.MigratedInvoiceTextSearchPages.Value);
+            }
+
+            // Refresh settings since these operations may run for very long time
+        }
+
+        private async Task MigratedInvoiceTextSearchToDb(int startFromPage)
+        {
+            using var ctx = _dbContextFactory.CreateContext();
```

Comment:
You should use a transaction to ensure all the data is saved without issue
```suggestion
            using var ctx = _dbContextFactory.CreateContext();
            using var transaction = ctx.Database.BeginTransaction();
```

Reply:
https://github.com/btcpayserver/btcpayserver/pull/2029#discussion_r549137609